### ISSUE TITLE
We need a good IPC mechanism.

### ISSUE THREAD

**Issue by Yoric** [redox-os/redox#839]
Having a good IPC mechanism is extremely useful.

By "good", I mean:
- high-level;
- easy on the kernel;
- fast;
- capabilities to send/receive should be file descriptors.

We might not get there in one step, but let's start the discussion on how we can get there.

**Comment by Yoric** [comment]
Let's start with "easy on the kernel":
- data should not go through the kernel, which means that we need shared memory;
- synchronization should not go through the kernel (in mosts cases), which means that we need futexes.

That's already two dependencies.

**Comment by DaanHoogland** [comment]
So we will have a process that provides mutexes and shared memory. I think that could or even should be one process

**Comment by rlfnb** [comment]
Did you look into the grant mechanism used by Minix3? http://wiki.minix3.org/doku.php?id=developersguide:memorygrants
Basically, I find it quite sophisticated and it allows some kind of shared memory as well.

**Comment by Yoric** [comment]
Now, let's look at "capabilities to send/receive should be file descriptors".

At first, I thought that it would be sufficient to have a `ipc:` scheme whose main role would be to open/close the communication channels, by (de)initializing the shared memory and futex and granting the process the right to use each.

However, I believe that this is not sufficient to ensure that a context stops using a shared memory segment after the `ipc` is closed.

**Comment by Yoric** [comment]
@rlfnb Adding to the task list.

**Comment by mark-i-m** [comment]
Have you considered something like [this](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/osr2007_rethinkingsoftwarestack.pdf)?

Of course, the implementation from Singularity won't work as-is in Redox, but it could be good inspiration... Singularity channels as described were intended to be fast (zero-copy), capability-based, and easy to reason about formally, which seems to be the requirement here, too.

Just an idea :)

**Comment by boricj** [comment]
Hi all, I'm an occasional MINIX contributor and I happen to know a bit on MINIX's memory grants behind the scenes. I'll just make a brain-dump here from the point of view of a kernel developer.

Memory grants are used to transfer arrays of bytes across virtual address spaces. Synchronization (along with process inter-communication) is done through the message passing IPC. For example, to get a directory listing from a FS server to the VFS server, VFS sends a message to FS with a memory grant in it, FS uses the grant to write the listing in VFS's address space and then sends back a reply message when it's done with the request. VFS then terminates the grant and carries on.

Memory grants aren't shared memory, system calls are required to copy memory around (which allows byte-granularity bounds to be enforced among other things). The micro-kernel itself copies directly from a virtual address space to another, without involving the VM server if there's no page faults. Tricks like indirect and magic grants along with shared memory buffer caches are used to keep the numbers of copies to a minimum.

Grants are extremely cheap to set up and tear down since the program merely tells the micro-kernel where in memory is its table of grants on startup and updates it as needed on the fly with no kernel calls. Along with 64 bytes message passing, it's enough for MINIX 3 to implement all UNIX IPCs (even shared memory, since it's just sending messages to the VM server in order to make magic happen) as well as nearly all the internal plumbing.

I guess if you really need the speed you can just use shared memory, but you'd at least lose byte granularity, fine-grained grant transfer and dirt-cheap set up. I wasn't around when memory grants were introduced, but I think they predated virtual memory support in MINIX.

The relevant micro-kernel code is at https://github.com/Stichting-MINIX-Research-Foundation/minix/blob/master/minix/kernel/system/do_safecopy.c. The paper linked at the bottom of the wiki page has a more technical in-depth explanation in 3 pages from the point of view of user programs.

In short, memory grants are :
- low-level (you get an array of bytes, the micro-kernel does not cares what it means),
- easy on the kernel as long as page faults do not occur,
- reasonably fast (really cheap to set up, VM isn't involved if there's no page fault, no in-kernel buffering) as long as you use tricks,
- safe,
- but can't send file descriptors by themselves,
- and they require some form of pre-existing IPC to set up.

**Comment by nosideeffects** [comment]
I've had a good experience using [Aeron's](https://github.com/real-logic/Aeron) IPC mode. Might be a good design to base a Redox IPC on.

**Comment by dathinab** [comment]
It might also be worth to take a glance at bus1: http://www.bus1.org/bus1.html
(it's linux specific, but should still be worth to take a look at it's design)
